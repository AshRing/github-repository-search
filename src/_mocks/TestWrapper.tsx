import * as React from "react";

import { MemoryRouter } from "react-router";
import { ThemeProvider } from "styled-components";
import { theme } from "../_style";

const TestWrapper = ({ children }: { children: React.ReactChild }) => (
	<ThemeProvider theme={theme}>
		<MemoryRouter initialEntries={[{ pathname: "/", key: "head-test-key" }]}>
			{children}
		</MemoryRouter>
	</ThemeProvider>
);

export { TestWrapper };
