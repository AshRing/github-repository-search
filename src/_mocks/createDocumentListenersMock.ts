// used to test useClickOutside hook w/i components
export const createDocumentListenersMock = () => {
    const listeners = {};
    const handler = (domEl, event) => listeners?.[event]?.({ target: domEl });
    document.addEventListener = jest.fn((event, cb) => {
        listeners[event] = cb;
    });

    document.removeEventListener = jest.fn((event) => {
        delete listeners[event];
    });

    return {
        mouseDown: (domEl) => handler(domEl, "mousedown"),
        click: (domEl) => handler(domEl, "click"),
    };
};
