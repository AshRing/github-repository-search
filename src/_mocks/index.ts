export * from "./repo";
export * from "./TestWrapper";
export * from "./createDocumentListenersMock";
