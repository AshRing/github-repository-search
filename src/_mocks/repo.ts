import { IRepository } from "../_types";

export const mockRepo1: IRepository = {
	created_at: "10/01/2021",
	description: "",
	forks_count: 123,
	full_name: "@ash/test-project",
	homepage: "",
	html_url: "https://github.com",
	id: "1",
	language: "JavaScript",
	name: "test-project",
	open_issues_count: 0,
	owner: {
		login: "ash",
		id: "123",
		avatar_url: "some_url",
		url: "some_url",
	},
	stargazers_count: 55,
	updated_at: "10/03/2021",
	watchers_count: 12,
};
