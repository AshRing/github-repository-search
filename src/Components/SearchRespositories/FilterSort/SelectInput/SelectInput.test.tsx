import * as React from "react";

import {
	ClearButton,
	DropdownArrow,
	SelectDropdown,
	SelectInputContainer,
	SelectOption,
} from "./SelectInput.styles";
import { TestWrapper, createDocumentListenersMock } from "../../../../_mocks";
import { mount, shallow } from "enzyme";

import { SelectInput } from ".";

let props;

describe("single select", () => {
	beforeEach(() => {
		props = {
			handleChange: jest.fn(),
			label: "Color",
			options: ["Red", "Green", "Blue"],
			selected: [],
			multiselect: false,
		};
	});

	test("Single select SelectInput matches snapshot", () => {
		const wrapper = shallow(<SelectInput {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("clicking the input container opens the select dropdown & has correct dropdown buttons", () => {
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		expect(wrapper.find(SelectDropdown).prop("open")).toBeFalsy();
		expect(wrapper.find(DropdownArrow).prop("dropdownOpen")).toBeFalsy();
		wrapper.find(SelectInputContainer).simulate("click");
		wrapper.update();
		expect(wrapper.find(SelectDropdown).prop("open")).toBeTruthy();
		expect(wrapper.find(DropdownArrow).prop("dropdownOpen")).toBeTruthy();
	});

	test("close icon closes select dropdown", () => {
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		wrapper.find(SelectInputContainer).simulate("click");
		wrapper.update();
		expect(wrapper.find(SelectDropdown).prop("open")).toBeTruthy();
		wrapper.find("#dropdownCloseIcon").simulate("click");
		wrapper.update();
		expect(wrapper.find(SelectDropdown).prop("open")).toBeFalsy();
	});

	test("single select doesn't display clear button", () => {
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		expect(wrapper.find(ClearButton).exists()).toBeFalsy();
	});

	test("clicking an option selects that option and closes dropdown", () => {
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		wrapper.find(SelectInputContainer).simulate("click");
		wrapper.update();
		expect(wrapper.find(SelectDropdown).prop("open")).toBeTruthy();
		const firstOption = wrapper.find(SelectOption).first();
		const optionText = firstOption.find("span").text();
		firstOption.simulate("click");
		wrapper.update();
		expect(props.handleChange).toHaveBeenCalledWith([optionText]);
		expect(wrapper.find(SelectDropdown).prop("open")).toBeFalsy();
	});

	test("click outside closes select dropdown", () => {
		const fireEvent = createDocumentListenersMock();
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		wrapper.find(SelectInputContainer).simulate("click");
		wrapper.update();
		expect(wrapper.find(SelectDropdown).prop("open")).toBeTruthy();
		fireEvent.mouseDown(document.body);
		wrapper.update();
		expect(wrapper.find(SelectDropdown).prop("open")).toBeFalsy();
	});
});

describe("multiselect", () => {
	beforeEach(() => {
		props = {
			handleChange: jest.fn(),
			label: "Color",
			options: ["Red", "Green", "Blue"],
			selected: ["Red"],
			multiselect: true,
		};
	});

	test("multiselect SelectInput matches snapshot", () => {
		const wrapper = shallow(<SelectInput {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("multiselect doesn't displays clear button if nothing is selected", () => {
		props = {
			...props,
			selected: [],
		};
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		expect(wrapper.find(ClearButton).exists()).toBeFalsy();
	});

	test("multiselect displays clear button if options are selected & clicking it clears selected options", () => {
		props = {
			...props,
			selected: ["Red"],
		};
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		expect(wrapper.find(ClearButton).exists()).toBeTruthy();
		wrapper.find(ClearButton).simulate("click");
		expect(props.handleChange).toHaveBeenCalledWith([]);
	});

	test("clicking a multiselect option selects it if it hasn't been selected", () => {
		const selectColor = "Green";
		props = {
			...props,
			selected: ["Red"],
		};
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		wrapper.find(SelectInputContainer).simulate("click");
		wrapper.update();
		expect(wrapper.find(SelectDropdown).prop("open")).toBeTruthy();
		wrapper
			.find(SelectOption)
			.findWhere((w) => w.text().includes(selectColor))
			.first()
			.simulate("click");
		expect(props.handleChange).toHaveBeenCalledWith([...props.selected, selectColor]);
	});

	test("clicking a multiselect option deselects option if it had been selected", () => {
		const selectColor = "Green";
		props = {
			...props,
			selected: [selectColor],
		};
		const wrapper = mount(
			<TestWrapper>
				<SelectInput {...props} />
			</TestWrapper>,
		);
		wrapper.find(SelectInputContainer).simulate("click");
		wrapper.update();
		expect(wrapper.find(SelectDropdown).prop("open")).toBeTruthy();
		wrapper
			.find(SelectOption)
			.findWhere((w) => w.text().includes(selectColor))
			.first()
			.simulate("click");
		const optionRemoved: string[] = props.selected.filter(
			(selectedOption: string) => selectedOption !== selectColor,
		);
		expect(props.handleChange).toHaveBeenCalledWith(optionRemoved);
	});
});
