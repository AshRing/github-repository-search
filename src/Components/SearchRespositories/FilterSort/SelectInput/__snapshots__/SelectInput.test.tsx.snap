// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiselect multiselect SelectInput matches snapshot 1`] = `ShallowWrapper {}`;

exports[`single select Single select SelectInput matches snapshot 1`] = `ShallowWrapper {}`;
